packages/ember-data/lib/system/model/internal-model.js

```diff
@@ -0,0 +1,696 @@
+import merge from "ember-data/system/merge";
+import RootState from "ember-data/system/model/states";
+import createRelationshipFor from "ember-data/system/relationships/state/create";
+import Snapshot from "ember-data/system/snapshot";
+import Errors from "ember-data/system/model/errors";
+
+var Promise = Ember.RSVP.Promise;
+var get = Ember.get;
+var set = Ember.set;
+var forEach = Ember.ArrayPolyfills.forEach;
+var map = Ember.ArrayPolyfills.map;
+
+var _extractPivotNameCache = Ember.create(null);
+var _splitOnDotCache = Ember.create(null);
+
+function splitOnDot(name) {
+  return _splitOnDotCache[name] || (
+    _splitOnDotCache[name] = name.split('.')
+  );
+}
+
+function extractPivotName(name) {
+  return _extractPivotNameCache[name] || (
+    _extractPivotNameCache[name] = splitOnDot(name)[0]
+  );
+}
+
+function retrieveFromCurrentState(key) {
+  return function() {
+    return get(this.currentState, key);
+  };
+}
+
+/*
+  `InternalModel` is the Model class that we use internally inside Ember Data to represent models.
+  Internal ED methods should only deal with `InternalModel` objects. It is a fast, plain Javascript class.
+
+  We expose `DS.Model` to application code, by materializing a `DS.Model` from `InternalModel` lazily, as
+  a performance optimization.
+
+  `InternalModel` should never be exposed to application code. At the boundaries of the system, in places
+  like `find`, `push`, etc. we convert between Models and InternalModels.
+
+  We need to make sure that the properties from `InternalModel` are correctly exposed/proxied on `Model`
+  if they are needed.
+*/
+
+var InternalModel = function(type, id, store, container, data) {
```

Comment:
`function InternalModel` so that cross platform we get good naming on stacks and debugging.